README.md
# BDOBot
This project is a Discord Bot written in Python for the online MMORPG Black Desert Online.
The bot is my first Discord bot and my first side project that I have decided to take up in order to create some convenience.

## Features
- This bot provides multiple features that may aid in the gameplay or additional fun features
- Gameplay related features: 
    - `%foodname`: retrieves ingredients for cooking / alchemy (i.e. %5 beer)
    - `%wb`: shows upcoming world boss and time
    - `%bosshunter`: register or unregister for a 'Boss Hunter' role to be notified of boss spawns
    - `%mp`: calculates profits for marketplace sales
    - `%hystria`: sends a map of Hystria dungeon
    - `%sycraia`: sends a map of Sycraia underwater dungeon
- Non gameplay related features:
    - `%popcorn`: register or unregister for a 'Caramel Popcorn' role to be notified when there's a movie stream in the guild
    - `%anime`: uses mal API to retrieve anime information from MyAnimeList
    - `%manga`: uses mal API to retrieve manga information from MyAnimeList
    - `%calc`: calculator
    - `%meme`: uses praw API to retrieve random memes from selected subreddits
- Additional features:
    - `%recruit`: updates recruitment status to determine newcomer message
    - Automatic message for when a new user joins, sending welcome and Q&A or recruitment closed message depending on %recruit status
    - `%khan`: updates Khan announcement status
    - Automatic Khan raid announcement for next upcoming Khan raid
    - `%war`: updates Node War announcement status
    - Automatic Node War announcement for next upcoming node war

## Future implementation
- Implementation of dockerized noSQL database to store configuration values


## Contribution
- Pull requests are welcomed, as well as suggestions or code critic or reviews. This bot is still in the development phase and will be for a while as it is a side project.
- Some of the code are working but not the most efficient or elegant, any pull request for cleaner code is always appreciated.


CHANGELOG.md
# Changelog

## 4 Nov 2021
- Removed ping and reactions on node war announcement message
- Added handling for %intro if user is unable to be fetched to send a generic message instead

## 21 Oct 2021
- Updated war announcement message to be 3 times a week, added new logic to retrieve date
- Removed snipe reminders

## 9 Sep 2021
- Removed use of secrets.txt and praw_secrets.txt, secrets now grabbed from .env
- Added Development Cog for new features to be developed in there
- Added development mode in Constants.py to toggle between production and development mode (dev mode disables all other modules live)

## 6 Sep 2021
- Fixed missing positional arguments if no snipe duty for the day

## 2 Sep 2021
- Fixed closed recruitment message line breaks
- Updated %snipe to send schedule if no .csv file is submitted

## 1 Sep 2021
- Fixed a bug where logging error causes snipe reminder to fail
- Fixed missing line in war announcement

## 31 Aug 2021
- Fixed missing Constants causing closed recruitment message to not send
- Added %intro feature to resend questionnaire
- Condensed all role giving features (%bh, %popcorn, %doughtart) into a single %role menu

## 30 Aug 2021
- Added check for if snipe duty officer exist to prevent null

## 26 Aug 2021
- Updated Welcome, Khan, Node War messages 

## 24 Aug 2021
- Added %halp as a %help alternative (@/Laviz request)
- Added %uwu to send random golden retriever pics (let me know if you have more subreddit for cute animals)
- Added %snipe to receive a valid .csv file of [DD-MM-YYYY,Name,Discord_ID,NotificationPreference,Day] format from discord uploads
- Added %duty to show officer on node war snipe duty for the day
- Added auto reminder to ping duty officer at 1730hrs GMT+8 in bot-channel

## 1 Aug 2021
- Added %doughtart to register/unregister dota role
- Added scheduled task daily to update 'Guild Member' roles for Croissant/Macaron/Eggtart

## 19 Jul 2021
- fixed error in %mp not showing correct profit values

## 16 Jul 2021
- fixed %meme with asyncpraw
- migrated bot to amazon AWS EC2 (again, finally)
- fixed a bug where khan / war announcement will be set to next announcement date if bot is restarted on announcement day, even if announcement has yet to be sent

## 9 Jul 2021
- added automated Khan and node war announcements
- added %khan and %war for officers to toggle announcement status
- updated welcome message to have poster on top and link to #about-us properly

## 16 May 2021
- added %sycraia
- default recruitment status to open

## 15 Apr 2021
- Updated sad message when recruitment is closed
- Added %recruit for officers to update guild recruitment status. if recruitment is closed bot will send a sad message instead of the whalecome message

## 14 Apr 2021
- Added %manga and %popcorn information in %help
- Added %popcorn to (un)register yourself to be part of the movie gang
- Added an error message to be sent when retrieving anime or manga results in an error

## 9 Apr 2021
- Increased delay for welcome message to 25 seconds
- Redid %anime
- Added %manga
- Backend cleanup

## 11 Jan 2020
- BETA: added %anime feature to search for anime and returns a synopsis and an image

---- Lost in time ----

## 12 Jan 2019
added %calc - calculator function 


Constants.py
# CONFIGS
## Set Bot MODE to DEV if developing. All Cogs will be disabled!
## DEVELOPMENT MODE
# MODE='DEV'
# PREFIX_PERCENT = '&'

## PRODUCTION MODE
MODE='PROD'
PREFIX_PERCENT = '%'

STATUS_LOOP_TIME_SECONDS = 20
REACTION_TIMEOUT_SECONDS = 7
MEME_SUBREDDITS = ['memes', 'animemes', 'meme', 'goodanimemes']
UWU_SUBREDDITS = ['goldenretrievers']
BOSS_NOTIFICATION_NOTICE_SECONDS = 1800
MODULES_GENERAL = 'Modules.General'
MODULES_MARKET = 'Modules.Market'
MODULES_BOSS = 'Modules.Boss'
MODULES_FUN = 'Modules.Fun'
MODULES_ANNOUNCE = 'Modules.Announcement'
MODULES_OFFICER = 'Modules.Officer'
MODULES_ROLE = 'Modules.Role'
MODULES_DEV = 'Modules.Development'
STARTUP_COG_MODULES = [MODULES_GENERAL, MODULES_MARKET, MODULES_BOSS, MODULES_FUN, MODULES_ANNOUNCE, MODULES_OFFICER, MODULES_ROLE] if MODE == 'PROD' else [MODULES_DEV]


# LOGGING
COG_STARTUP = "starting up %s Cog"
ON_MEMBER_JOIN = "%s has joined, sleeping for 25 seconds before sending welcome"
ON_MEMBER_REMOVE = "%s has left or have been removed from the server"
BOT_STARTUP = "starting up bot with ID: %s, Name: %s"
RECRUIT_TRIGGER = "Recruitment: %s triggered by %s"
KHAN_TRIGGER = "Khan: %s triggered by %s"
WAR_TRIGGER = "Node War: %s triggered by %s"
REACTION_CHOSEN = "reaction chosen: %s"
REACT_TIMEOUT = "%s reaction timeout hit, removing message"
LOGGING_STREAM_FORMAT = '[LOG] [%(asctime)s] [%(levelname)s] [%(funcName)s:%(lineno)s] : %(message)s'
EXT_FAILED_TO_RUN = "an extension failed to run - {0}: {1}"
SCHEDULER_STARTUP = "starting up %s scheduler"
NEXT_ANNC = "next %s announcement set at %s"
SENT_ANNC = "sent %s announcement at %s"
UPCOMING_ANNC = "upcoming %s announced happening at %s"
INACTIVE_ANNOUNCER = "%s announcement INACTIVE, no message sent"
FUN_QUERY_SENT = "%s name passed in: %s"
FUN_QUERY_ERR = "unable to retrieve %s name: %s"
BOSS_NOTIFICATION_SENT = "sending boss notification for %s"
RETRIEVE_PRAW_CONFIG = "retrieving PRAW ID and secret"
ENV_VAR_NOT_FOUND = "%s env var not found, attempting to read %s"
FILE_NOT_FOUND = "%s not found"
RETRIEVE_DISCORD_CONFIG = "retrieving Discord token"
CONFIG_FILE_RETRIEVE_SUCCESS = "%s params retrieved from %s"
NEXT_ROLE_UPDATE_TIME = "next roles update time set at %s"
ROLE_UPDATED_TIME = "roles updated at %s"
NO_ROLE_UPDATED = "no %s role updated today"
ROLE_UPDATED = "Following %s have %s role %s in their roles: [%s]"
ADD_REACTION_FOR_MSG = 'message passed and reaction type %s added'

# NUMBER / FORMATS
ZERO = 0
ONE = 1
TWO = 2
TEN = 10
EIGHTEEN = 18
TWENTY_ONE = 21
SIXTEEN = 16
THIRTY = 30
DT_FORMAT_ANNC = '%d/%m/%Y, %H:%M:%S'
DT_FORMAT_INVITE = 'dddd DD MMM YYYY hh:mm A'
DT_FORMAT_WB = '%H:%M'
DT_FORMAT_SNIPE = 'DD-MM-YYYY'

#COMMANDS
HELP_L = 'help'
BUG_L = 'bug'
HYSTRIA_L = 'hystria'
SYCRAIA_L = 'sycraia'
CALC_L = 'calc'
RECRUIT_L = 'recruit'
HELLO_L = 'hello'
NEZUKO_L = 'nezuko'
KHAN_L = "khan"
TWITCHTV_L = 'twitch.tv'
WAR_L = "war"
MEME_L = "meme"
ANIME_L = "anime"
MANGA_L = "manga"
POPCORN_L = "popcorn"
MARKETPLACE_L = "mp"
WB_L = "wb"
BOSSHUNTER_L = "bosshunter"
DOUGHTART_L = "doughtart"
HALP_L = 'halp'
UWU_L = 'uwu'
SNIPE_L = "snipe"
DUTY_L = "duty"
INTRO_L = "intro"
ROLE_L = "role"

# STRING
PASTRIES = "Pastries"
NEW_LINE = '\n'
RECRUIT = "Recruit"
KHAN = "Khan"
NODE_WAR = "Node War"
KHAN_ANNC = "Khan announcement"
WAR_ANNC = "War announcement"
STATUS = "STATUS"
OPEN = "OPEN"
CLOSE = "CLOSE"
ACTIVE = "ACTIVE"
INACTIVE = "INACTIVE"
STATUS_MSG_USE_HELP = "use %help"
NEZUKO_BOT = "NezukoBot"
ANIME = "Anime"
MANGA = "Manga"
CARAMEL_POPCORN = "Caramel Popcorn"
GENERAL = "General"
ANNOUNCEMENT = "Announcement"
FUN = "Fun"
MARKET = "Market"
BOSS = "Boss"
BOSS_HUNTER = "Boss Hunter"
PRAW_CLIENT_ID = "PRAW_CLIENT_ID"
PRAW_CLIENT_SECRET = "PRAW_CLIENT_SECRET"
DISCORD_TOKEN = "DISCORD_TOKEN"
DOTENV_FILE = ".env"
FILE_READ_MODE = 'r'
ROLE_UPDATE = "role update"
CROISSANT = "Croissant"
MACARON = "Macaron"
EGGTART = "Eggtart"
GUILD_MEMBER = "Guild Member"
ADDED = "added"
REMOVED = "removed"
DOUGHTART = "DO-ugh-TArt"
DND_CRUMBS = "D&D Crumbs"
KHAN_CAPS = 'KHAN'
WAR_CAPS = 'WAR'
UPDATE = 'UPDATE'
YES_NO = 'YES_NO'
ROLES_UPDATE_COMD = "ROLES_UPDATE"
SNIPE_SCHEDULE_FILE = "snipe_schedule_file.csv"
SNIPE_REMINDER = "Snipe Reminder"
OFFICER = "Officer"
ROLE = "Role"


# ANIME/MANGA
MSG_FUN_RESPONSE = '```{0}```\n```{1}```\n{2}'
MSG_FUN_ERR = "An error occured while retrieving {0} {1} :("

# BOSS
MSG_NEXT_BOSS_ANNC = "```diff\n- The next boss is {0} at {1} hrs, GMT+8```"
MSG_NEXT_BOSS_REMINDER = "{0} \n```md\n {1} will spawn in 30 minutes time!```"

# CALC
MSG_CALC_FAIL = "```Sorry, I didn\'t get that, please use +, -, *, /, % only!```"
MSG_CALC_RES = "```{0} is {1}```"

# GENERAL
MSG_BUG_DISCOVERED = "<@!{}> A bug have been discovered!"
MSG_COMD_DENIED = "```Only Crème brûlées are allowed to use this feature!```"
MSG_HELLO_MENTION = "Hello {0.mention}!"
MSG_HELP_MSG = "```yaml\n" \
          "List of Commands\n----------------\n\n" \
          "%help - shows this message\n\n" \
          "%halp - shows this message\n\n" \
          "%foodname - retrieves ingredients i.e. %5 beer\n\n" \
          "%role - displays the role menu to register / unregister yourself for custom roles \n\n" \
          "%wb - shows the upcoming world boss \n\n" \
          "%mp - shows profit with(out) VP from selling in MP i.e. %mp 43500000\n\n" \
          "%hystria - sends a map of hystria!\n\n" \
          "%meme - sends a random meme\n\n" \
          "%uwu - sends a golden retriever doggo\n\n" \
          "%calc - built-in calculator, supports +, -, *, /, %\n\n" \
          "%anime <title> - retrieves anime information from MyAnimeList\n\n" \
          "%manga <title> - retrieves manga information from MyAnimeList\n\n" \
          "%intro <@person> - Sends questionnaire (Brioche Buns only!)\n\n" \
          "%recruit - Update recruitment status (Crème brûlées only!)\n\n" \
          "%khan - Update Khan announcement status (Crème brûlées only!)\n\n" \
          "%war - Update Node War announcement status (Crème brûlées only!)\n\n" \
          "%snipe - Upload a .csv file for daily node war snipe reminder (Crème brûlées only!)\n\n" \
          "%bug <message> - reports a bug to Kagi\n\n```"
MSG_INVALID_COMD = "No results or commands was retrieved." \
        "\nPerhaps you wanna use '%help'?"
MSG_ON_MEMBER_REMOVE = "```{0} has left or have been removed from the server```"
MSG_TWITCH_WARNING = "Hello {0.mention}, it seems like you have pasted a Twitch link. All videos or stream advertisements goes to {1.mention}, thank you!"
MSG_ROLE_REGISTER = "{0.mention} is now registered as a {1}!"
MSG_ROLE_RESIGN = "{0.mention} have resigned as a {1}!"
MSG_ROLE_UPDATE = "Following `{0}` have `{1}` {2} in their roles: \n```{3}```"

# MARKETPLACE
MSG_MP_RESULT = "Without VP: {0} \nWith VP: {1}\nValues might be a little off depending on your family fame!"

# KHAN
MSG_KHAN_ENQUIRY = 'Please react to enquire or update Khan announcement status!\n\n' \
                ':bread: - Khan Announcement Status\n\n:o: - Activate Khan Announcement \n\n:x: - Deactivate Khan Announcement'
MSG_KHAN_OPEN_UPDATE = "{0.mention} ```Updated Khan Announcement status to active!```"
MSG_KHAN_CLOSE_UPDATE = "{0.mention} ```Updated Khan Announcement status to inactive!```"
MSG_KHAN_STATUS = "```Khan Announcement Status: {0}```"

MSG_KHAN_INVITE = ':whale:**HI BREADS! LET\'S DO KHAN RAID TOMOROWWW!!:whale:\n\n' \
':calendar_spiral:Mark your calendar!**\n' \
'`{0} GMT+8`\n\n' \
':pretzel:Server will be announced in guild chat 1 hour before the raid starts!\n' \
':pretzel:Please bring your PVP-safe character (Lv49 or less) to avoid grifers!\n' \
':pretzel:CTG to OE, and taxi back to mainland will be provided\n' \
':pretzel:We will be doing XL (Tier 3)! Let us know if you have friends interested to come to WH!\n\n' \
'> *New to Khan Raid? Just come and join uss! We fully support every breads who wanna try new experiences! <:yay:707879651804053565>\n' \
'> No gear req needed as you\'ll shoot em fish with cannon! >:D*\n' \
'> Check this Video Guide to learn about Cannon Rotation so you can deal damage to the big fish more efficiently!\n' \
'> https://www.youtube.com/watch?v=h1jY7r6bGQk\n' \
'> Or kindly poke {1.mention} for your Khan related questions!\n\n' \
'**Confirm your attendance**\n' \
'> by reacting :regional_indicator_y: if you\'re going to attend, or :regional_indicator_n: if you\'re not!\n' \
'*Don\'t miss out seeing El getting Khan\'s Heart! See you this Saturday, breados!*<:yay:707879651804053565>' \
'<@&635776028048097290> <@&574641855229067264>'

# RECRUIT
MSG_REC_CLOSED_MSG = ':pretzel:**Welcome to Pastries!**:pretzel:\n\n' \
                'Hello {0.mention}, Thank you for visiting our discord!\n' \
                'We\'d really love to have you joining our little bakery.\nBut unfortunately, ' \
                ':no_entry_sign:**our recruitment is CLOSED**:no_entry_sign: for the time being!\n' \
                'Hopefully we\'d meet again in the next chance!\n\nIf you have any questions ' \
                'feel free to mention @/[Officer]Brioche Bun!\n' \
                '*While you\'re here, please have some free cheese garlic bread* :sparkles:\n'
MSG_RECRUIT_ENQUIRY = "Please react to enquire or update recruitment status!\n\n" \
    ":bread: - Recruitment Status\n\n:o: - Opening of recruitment \n\n:x: - Closing of recruitment"
MSG_RECRUIT_OPEN_UPDATE = "{0.mention} ```Updated recruitment status to open!```"
MSG_RECRUIT_CLOSE_UPDATE = "{0.mention} ```Updated recruitment status to closed!```"
MSG_RECRUIT_STATUS = "```Recruitment Status: {0}```"

MSG_REC_OPEN_MSG =':sparkles::pretzel:**Hello fresh dough {0}, welcome to Pastries!**:sparkles::pretzel:\n' \
               'We would like you to answer the following short questions before proceeding!\n\n' \
               '> **01. What\'s your in-game Family Name? Where did you find our guild from?** *i.e. from server-chat advertisement, recommendation from a friend (please identify), Guild Rank Board, BDO SEA community forum, etc*\n\n' \
               '> **02. What are you looking for in a guild?** *i.e. Payouts, socializing, guidance in game*\n\n' \
               '> **03. Have you been in a guild before?** *If YES, what guild and why did you leave?*\n\n' \
               '> **04. What kind of player do you consider yourself as?** *PVE, PVP, Lifeskill-oriented, or an all-rounded?*\n\n' \
               '> **05. OPTIONAL - Where are you from? What timezone do you live in?** *We got people from MY/SG/INA/PH and a few more other countries inside!*\n\n' \
               ':cake:By answering these questions, you also acknowledge that Pastries is a **chill International PVX/Lifeskill guild**\n' \
               ':cake:**We do not tolerate any form of unhealthy or immature behaviours when you are with us!**\n' \
               ':cake:Please head over to {1.mention} for more info and our guild rules!\n\n' \
               '**Thank you for answering the questions - Please mention @/[Officer]Brioche Bun when you are done**\n' \
               '*We look forward to get to know you better!*\n\n'

# NODE WAR
MSG_WAR_ENQUIRY = 'Please react to enquire or update NW announcement status!\n\n' \
                ':bread: - NW Announcement Status\n\n:o: - Activate NW Announcement \n\n:x: - Deactivate NW Announcement'
MSG_WAR_OPEN_UPDATE = "{0.mention} ```Updated War Announcement status to active!```"
MSG_WAR_CLOSE_UPDATE = "{0.mention} ```Updated War Announcement status to inactive!```"
MSG_WAR_STATUS = "```War Announcement Status: {0}```"
MSG_NO_ATTACHMENTS = "```No attachments found in message, please send .csv with valid format!```"


MSG_WAR_INVITE = ':crossed_swords:**PREPARE YOUR KNAIFU WE GOT SOME BUTTER TO SPREAD!:crossed_swords:\n\n' \
':calendar_spiral:Mark your calendar!**\n' \
'`{0} GMT+8`\n`{1} GMT+8`\n`{2} GMT+8`\n\n' \
':pretzel:We will be participating in **Tier 1** Node War @ **Serendia 1**\n' \
':pretzel:**We need as many participants as possible!** No minimum gearscore required, EVERYONE is invited to join in the fun, the more the merrier!<:yay:707879651804053565>\n' \
':pretzel:Shotcalls will be on Discord so joining in voice chat is mandatory! <:kagi_proudmum:830388637073145907>\n\n' \
'> *New to Node War? Just come and join us! We fully support every breads who wanna try new experiences!*<:yay:707879651804053565>\n' \
'> While knowing how to PVP is crucial, there are so much more into NW if you prefer the \'less-agressive\' way to contribute, i.e. Carrack-captain, Elephant-rider + Spear-thrower, Flame Tower pilot, Cannoneer, Hwacha pilot, etc!\n' \
'> Kindly poke any of our officers if you\'re interested or have any questions in any of these stuff!\n' \
'> **Pastres\' NW Pocket guide** Google Doc for our first-timer NW breads: https://tinyurl.com/6kshnxm3\n' \
'> If you have any questions feel free to ask in {3.mention}\n\n' \
'> **Please post/update your character gear info in {4.mention} to help us in making team compositions!**\n' \
'> More details about the node we will be fighting in and platoon composition will be posted here on war day!\n\n' \
'**CONFIRM YOUR ATTENDANCE**\n' \
'> by reacting on the days you are available to join us in {5.mention}!\n' \
'> All attendance announcements will be made 3 days before the war!\n'

# ROLE
MSG_ROLE_MENU = "Hello {0.mention}, welcome to the Roles menu!\n\n" \
        "In this menu, you can register / unregister for custom roles we have in Pastries\n" \
        "To update your roles, simply click on any of the reaction buttons provided -\n\n" \
        "> {1} - Caramel Popcorn - movie streaming time!\n" \
        "> {2} - DOugh-TArt - be part of the Dota gang!\n" \
        "> {3} - D&D Crumbs - join us and play D&D together!\n" \
        "> {4} - Boss Hunter - get notified of world boss 30 minutes before spawn!\n\n"

# ASSETS
ASSET_POSTER = "assets/poster.png"
ASSET_REC_CLOSE = "assets/garlic_bread.png"
ASSET_HYSTRIA = "assets/hystria.png"
ASSET_SYCRAIA = "assets/sycraia.png"
ASSET_NEZUKO = "assets/nezuko.gif"
ASSET_KHAN_ANNC = "assets/KHAN_ANN.png"
ASSET_NW_ANNC = "assets/NW_ANN.png"

# REACTION / EMOJI
EMOJI_STATUS = "\U0001f35e"
EMOJI_OPEN = "\U00002b55"
EMOJI_CLOSE = "\U0000274c"
EMOJI_Y = "\U0001F1FE"
EMOJI_N = '\U0001F1F3'
EMOJI_EL_YAY_C = '<:yay:707879651804053565>'
EMOJI_BAM_MAD_C = '<:bam_madman:868057992170463243>'
EMOJI_EL_SIP_C = '<:melvin_sip:867338932676329472>'
EMOJI_MIIA_GASP_C = '<:miia_gasp:868001842011992124>'
REACTION_ADD = 'reaction_add'
EMOJI_EL_YAY_C_ID = 707879651804053565
EMOJI_BAM_MAD_C_ID = 868057992170463243
EMOJI_EL_SIP_C_ID = 867338932676329472
EMOJI_MIIA_GASP_C_ID = 868001842011992124
EMOJI_REGIONAL_S = "\U0001F1F8"
EMOJI_REGIONAL_T = '\U0001F1F9'
EMOJI_REGIONAL_F = '\U0001F1EB'

# ID
ID_CHN_ABOUT_US = 575282309947850822
ID_CHN_ENTRY = 574641679575941137
ID_CHN_BAKERY = 574641537548550156
ID_CHN_MEDIA = 587559616188121108
ID_CHN_BOT_CHN = 574646616489721887
ID_CHN_MENU_BOARD =574650178183626771
ID_CHN_ACTIV_DISCUSS = 645931876107943967
ID_CHN_GEAR_CLASS = 801731373147881515
ID_CHN_KAGI_PLAYGROUND = 871250117120909362
ID_CHN_WAR_ATTENDANCE = 905295084797444127
ID_ROLE_BOSSHUNTER = 623100796833234955
ID_ROLE_BRIOCHE = 635776028048097290
ID_ROLE_CREME = 574641817505497106
ID_ROLE_CROISSANT = 574641855229067264
ID_ROLE_EGGTART = 659629130375102465
ID_ROLE_MACARON = 575188744617852939
ID_ROLE_POPCORN = 707958341539856404
ID_ROLE_GMEMBER = 801718825128034335
ID_ROLE_OFFICER = 801808081398792203
ID_ROLE_GM = 574802564596367361
ID_ROLE_DND_CRUMBS = 798722709739405332
ID_USER_EL = 193314826779361281
ID_USER_KAGI = 382152478810046464
ID_SERVER_PASTRIES = 574641536214630401
